File: backend/src/services/analytics.ts
import { Decimal } from "decimal.js";

import type { EmployeeRepository } from "../database/repos/employees.js";
import type { SalaryRepository } from "../database/repos/salaries.js";
import { DomainError } from "../common/errors.js";
import type { Employee, SalaryRecord } from "../common/types.js";
import type { FxService } from "./fx.js";

const GROUP_FIELDS = {
  country: "countryCode",
  department: "department",
} as const;

export class AnalyticsService {
  constructor(
    private readonly employees: EmployeeRepository,
    private readonly salaries: SalaryRepository,
    private readonly fx: FxService,
  ) {}

  private async groupActiveEmployees(groupBy: string) {
    if (!(groupBy in GROUP_FIELDS)) {
      throw new DomainError("groupBy must be country or department");
    }
    const field = GROUP_FIELDS[groupBy as keyof typeof GROUP_FIELDS];
    const buckets = new Map<string, Employee[]>();
    for (const employee of await this.employees.listActive()) {
      const list = buckets.get(employee[field]) ?? [];
      list.push(employee);
      buckets.set(employee[field], list);
    }
    return [...buckets.entries()].sort(([a], [b]) => a.localeCompare(b));
  }

  private async loadGroupedPay(groupBy: string, today: string) {
    const [buckets, rates] = await Promise.all([this.groupActiveEmployees(groupBy), this.fx.getRateTable()]);
    const currentByEmployeeId = await this.salaries.getCurrentForMany(
      buckets.flatMap(([, employees]) => employees.map((employee) => employee.id)),
      today,
    );
    return { buckets, rates, currentByEmployeeId };
  }

  private amountInUsd(record: SalaryRecord, rates: Map<string, Decimal>) {
    return this.fx.toUsd(record.baseAmount, record.currency, rates);
  }

  async headcount(groupBy = "country") {
    const buckets = await this.groupActiveEmployees(groupBy);
    return buckets.map(([group, employees]) => ({ group, headcount: employees.length }));
  }

  async spend(groupBy = "country", today: string) {
    const { buckets, rates, currentByEmployeeId } = await this.loadGroupedPay(groupBy, today);
    return buckets.map(([group, employees]) => {
      let spendUsd = new Decimal(0);
      let paidHeadcount = 0;
      for (const employee of employees) {
        const record = currentByEmployeeId.get(employee.id);
        if (!record) continue;
        spendUsd = spendUsd.add(this.amountInUsd(record, rates));
        paidHeadcount += 1;
      }
      return { group, spendUsd, paidHeadcount };
    });
  }

  async avgSalary(groupBy = "country", today: string) {
    const { buckets, rates, currentByEmployeeId } = await this.loadGroupedPay(groupBy, today);
    return buckets.map(([group, employees]) => {
      const paid = employees.flatMap((employee) => {
        const record = currentByEmployeeId.get(employee.id);
        return record ? [this.amountInUsd(record, rates)] : [];
      });
      const avgSalaryUsd =
        paid.length === 0
          ? null
          : paid.reduce((sum, value) => sum.add(value), new Decimal(0)).div(paid.length).toDecimalPlaces(2);
      return { group, avgSalaryUsd, paidHeadcount: paid.length, headcount: employees.length };
    });
  }
}

File: backend-py/app/services/analytics.py
from __future__ import annotations

from decimal import ROUND_HALF_UP, Decimal

from app.errors import DomainError
from app.models import Employee
from app.repos import EmployeeRepository, SalaryRepository
from app.services.fx import FxService

GROUP_FIELDS = {
    "country": "country_code",
    "department": "department",
}


class AnalyticsService:
    def __init__(
        self,
        employees: EmployeeRepository,
        salaries: SalaryRepository,
        fx: FxService,
    ):
        self.employees = employees
        self.salaries = salaries
        self.fx = fx

    def _group_active(self, group_by: str) -> list[tuple[str, list[Employee]]]:
        if group_by not in GROUP_FIELDS:
            raise DomainError("groupBy must be country or department")
        field = GROUP_FIELDS[group_by]
        buckets: dict[str, list[Employee]] = {}
        for employee in self.employees.list_active():
            key = getattr(employee, field)
            buckets.setdefault(key, []).append(employee)
        return sorted(buckets.items(), key=lambda item: item[0])

    def _load_grouped_pay(self, group_by: str, today: str):
        buckets = self._group_active(group_by)
        rates = self.fx.get_rate_table()
        ids = [employee.id for _, people in buckets for employee in people]
        current = self.salaries.get_current_for_many(ids, today)
        return buckets, rates, current

    def headcount(self, group_by: str = "country"):
        return [{"group": group, "headcount": len(people)} for group, people in self._group_active(group_by)]

    def spend(self, group_by: str, today: str):
        buckets, rates, current = self._load_grouped_pay(group_by, today)
        rows = []
        for group, people in buckets:
            spend_usd = Decimal("0")
            paid = 0
            for employee in people:
                record = current.get(employee.id)
                if not record:
                    continue
                spend_usd += self.fx.to_usd(record.base_amount, record.currency, rates)
                paid += 1
            rows.append({"group": group, "spend_usd": spend_usd, "paid_headcount": paid})
        return rows

    def avg_salary(self, group_by: str, today: str):
        buckets, rates, current = self._load_grouped_pay(group_by, today)
        rows = []
        for group, people in buckets:
            paid_amounts = []
            for employee in people:
                record = current.get(employee.id)
                if record:
                    paid_amounts.append(self.fx.to_usd(record.base_amount, record.currency, rates))
            if not paid_amounts:
                avg = None
            else:
                total = sum(paid_amounts, Decimal("0"))
                avg = (total / Decimal(len(paid_amounts))).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
            rows.append(
                {
                    "group": group,
                    "avg_salary_usd": avg,
                    "paid_headcount": len(paid_amounts),
                    "headcount": len(people),
                }
            )
        return rows
